File: Java/Observer/Calculator/src/com/ericgtkb/Observer.java
package com.ericgtkb;

public interface Observer {
    void update();
}

interface Display {
    void display();
}

class Add implements Observer, Display {
    private Subject subject;
    private int a;
    private int b;
    private int sum;

    public Add(Subject subject) {
        this.subject = subject;
        subject.addObserver(this);
    }

    private void calculateSum() {
        sum = a + b;
    }

    @Override
    public void update() {
        DataSource dataSource = (DataSource) subject;
        a = dataSource.getA();
        b = dataSource.getB();
        calculateSum();
        display();
    }

    @Override
    public void display() {
        System.out.println("Calculating sum...");
        System.out.println("a = " + a);
        System.out.println("b = " + b);
        System.out.println("a + b = " + sum);
    }
}

class Average implements Observer, Display {
    private Subject subject;
    private double aAve;
    private double bAve;
    private int count;

    public Average(Subject subject) {
        this.subject = subject;
        subject.addObserver(this);
    }

    private void calculateAverage(int a, int b) {
        aAve = (aAve * count + a) / (count + 1);
        bAve = (bAve * count + b) / (count + 1);
        count++;
    }

    @Override
    public void update() {
        DataSource dataSource = (DataSource) subject;
        calculateAverage(dataSource.getA(), dataSource.getB());
        display();
    }

    @Override
    public void display() {
        System.out.println("Calculating average...");
        System.out.println("The average of a so far is: " + aAve);
        System.out.println("The average of b so far is: " + bAve);
    }
}

File: C++/Observer/Calculator/observer.cpp
//
// Created by Eric Chen on 2019-09-05.
//

#include <iostream>
#include "observer.h"

Add::Add(Subject& s) : subject{&s}, a{}, b{}, sum{} {
    s.add_observer(*this);
}

void Add::update() {
    DataSource* data_source {dynamic_cast<DataSource*>(subject)};
    a = data_source->get_a();
    b = data_source->get_b();
    calculate_sum();
    display();
}

void Add::calculate_sum() {
    sum = a + b;
}

void Add::display() {
    std::cout << "Calculating sum..."
              << "\na = " << a
              << "\nb = " << b
              << "\na + b = " << sum << '\n';
}


Average::Average(Subject& s) : subject{&s}, a_ave{}, b_ave{}, count{0} {
    s.add_observer(*this);
}

void Average::update() {
    DataSource* data_source {dynamic_cast<DataSource*>(subject)};
    calculate_average(data_source->get_a(), data_source->get_b());
    display();
}

void Average::display() {
    std::cout << "Calculating average..."
              << "\nThe average of a so far is: " << a_ave
              << "\nThe average of b so far is: " << b_ave << '\n';
}

void Average::calculate_average(int a, int b) {
    a_ave = (a_ave * count + a) / (count + 1);
    b_ave = (b_ave * count + b) / (count + 1);
    count++;
}
